[JSC] Introduce JSPromiseReaction object

Constellation opened the pull request:
#### a1cb5e087a461a8ff468d49e85897349a175a2dd
<pre>
[JSC] Introduce JSPromiseReaction object
<a href="https://bugs.webkit.org/show_bug.cgi?id=299707">https://bugs.webkit.org/show_bug.cgi?id=299707</a>
<a href="https://rdar.apple.com/161528913">rdar://161528913</a>

Reviewed by Sosuke Suzuki.

This patch introduces JSPromiseReaction object. This is the first step
for cleaning up our Promise implementation mainly focusing on improving
JetStream3/doxbee-promise.

1. We simplify our reaction mechanism. Previously we were using
   array-likes, but in most of cases (1) user do not set multiple
   reactions to the same promise and (2) array tends to allocate more
   memory due to 2x resizing. So let&apos;s just use single-linked list of
   JSPromiseReaction.
2. JSPromiseReaction is internal JSObject, which has fixed slots
   accessible from JS side as well. Also DFG / FTL optimized allocation
   is set up.
3. Change of @pushNewPromiseReaction made @performPromiseThen bytecode
   size increased. And we observed large performance regression in
   JetStream3/doxbee-promise due to this. We annotate @performPromiseThen
   with @alwaysInline not to introduce this performance regression.

* JSTests/stress/destructuring-assignment-require-object-coercible-symbol-private.js:
(throw.new.Error):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.triggerPromiseReactions):
(linkTimeConstant.promiseResolveThenableJobFast):
(linkTimeConstant.promiseResolveThenableJobWithoutPromiseFast):
(linkTimeConstant.alwaysInline.performPromiseThen):
(linkTimeConstant.pushNewPromiseReaction): Deleted.
(linkTimeConstant.performPromiseThen): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsPromiseReaction):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::promiseReactionInternalFieldIndex):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseReactionInternalField):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseReactionInternalField):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNewInternalFieldObject):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseReactionStructure const):
* Source/JavaScriptCore/runtime/JSPromiseReaction.cpp: Added.
(JSC::JSPromiseReaction::createWithInitialValues):
(JSC::JSPromiseReaction::create):
(JSC::JSPromiseReaction::visitChildrenImpl):
(JSC::JSPromiseReaction::createStructure):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromiseReaction.h: Added.
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:

Canonical link: <a href="https://commits.webkit.org/300670@main">https://commits.webkit.org/300670@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f31669399a1a0d6dbbb64b3fe86ae784bdf43e19

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33864 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130171 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75590 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/62d16036-4587-4013-b876-b94fc4b51abf) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51762 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93840 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b0490852-37fb-44ee-8439-e8eaf232e41e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34973 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110451 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74471 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/13f3e83f-b6c6-4a99-940a-c1027ca7114c) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28609 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73688 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115612 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104688 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28834 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132890 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121985 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38367 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102334 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50779 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102186 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47544 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25771 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47208 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19439 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50258 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56019 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/152339 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49731 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38958 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53079 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51407 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4538,6 +4539,30 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
         }
 
+        case PromiseReactionCreateIntrinsic: {
+#if 1
```
Remove this directive?

sosukesuzuki commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -465,13 +466,8 @@ void Interpreter::getAsyncStackTrace(JSCell* owner, Vector<StackFrame>& results,
-                if (contextOffset != invalidOffset && !(contextFieldAttributes & (PropertyAttribute::Accessor | PropertyAttribute::CustomAccessorOrValue)))
-                    return reactions->getDirect(contextOffset);
-            }
+            if (auto* reaction = jsDynamicCast<JSPromiseReaction*>(reactionsValue))
```
Thanks!

sosukesuzuki reviewed (approved):
r=me but found typo in commit message `mechansim` -> `mechanism`

Constellation committed:
[JSC] Introduce JSPromiseReaction object
https://bugs.webkit.org/show_bug.cgi?id=299707
rdar://161528913

Reviewed by Sosuke Suzuki.

This patch introduces JSPromiseReaction object. This is the first step
for cleaning up our Promise implementation mainly focusing on improving
JetStream3/doxbee-promise.

1. We simplify our reaction mechanism. Previously we were using
   array-likes, but in most of cases (1) user do not set multiple
   reactions to the same promise and (2) array tends to allocate more
   memory due to 2x resizing. So let's just use single-linked list of
   JSPromiseReaction.
2. JSPromiseReaction is internal JSObject, which has fixed slots
   accessible from JS side as well. Also DFG / FTL optimized allocation
   is set up.
3. Change of @pushNewPromiseReaction made @performPromiseThen bytecode
   size increased. And we observed large performance regression in
   JetStream3/doxbee-promise due to this. We annotate @performPromiseThen
   with @alwaysInline not to introduce this performance regression.

* JSTests/stress/destructuring-assignment-require-object-coercible-symbol-private.js:
(throw.new.Error):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.triggerPromiseReactions):
(linkTimeConstant.promiseResolveThenableJobFast):
(linkTimeConstant.promiseResolveThenableJobWithoutPromiseFast):
(linkTimeConstant.alwaysInline.performPromiseThen):
(linkTimeConstant.pushNewPromiseReaction): Deleted.
(linkTimeConstant.performPromiseThen): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsPromiseReaction):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::promiseReactionInternalFieldIndex):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseReactionInternalField):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseReactionInternalField):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNewInternalFieldObject):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseReactionStructure const):
* Source/JavaScriptCore/runtime/JSPromiseReaction.cpp: Added.
(JSC::JSPromiseReaction::createWithInitialValues):
(JSC::JSPromiseReaction::create):
(JSC::JSPromiseReaction::visitChildrenImpl):
(JSC::JSPromiseReaction::createStructure):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromiseReaction.h: Added.
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:

Canonical link: https://commits.webkit.org/300670@main
